src/hypothesis/searchstrategy/types.py

```diff
@@ -0,0 +1,357 @@
+        strats = [from_type(hints[arg]) for arg in required & set(hints)]
+        if len(required) == sum(not v.is_empty for v in strats):
+            return st.builds(thing)
+    return st.nothing()
+
+
+def type_strategy_mapping(lookup=None):
+    """Return a dict mapping from types to corresponding search strategies.
+
+    Most resolutions will terminate here or in the special handling for
+    generics from the typing module.
+
+    """
+    import uuid
+    import decimal
+    import datetime as dt
+    import fractions
+
+    known_type_strats = {
+        type(None): st.none(),
+        bool: st.booleans(),
+        float: st.floats(),
+        complex: st.complex_numbers(),
+        fractions.Fraction: st.fractions(),
+        decimal.Decimal: st.decimals(),
+        text_type: st.text(),
+        bytes: st.binary(),
+        dt.datetime: st.datetimes(),
+        dt.date: st.dates(),
+        dt.time: st.times(),
+        dt.timedelta: st.timedeltas(),
+        uuid.UUID: st.uuids(),
+    }
+    for t in integer_types:
+        known_type_strats[t] = st.integers()
+    # build empty collections, as only generics know their contents
+    known_type_strats.update({
+        t: st.builds(t) for t in (tuple, list, set, frozenset, dict)
+    })
+    try:
+        from hypothesis.extra.pytz import timezones
+        known_type_strats[dt.tzinfo] = timezones()
+    except ImportError:  # pragma: no cover
+        pass
+    try:  # pragma: no cover
+        import numpy as np
+        from hypothesis.extra.numpy import \
+            arrays, array_shapes, scalar_dtypes, nested_dtypes
+        known_type_strats[np.dtype] = nested_dtypes()
+        known_type_strats[np.ndarray] = arrays(
+            scalar_dtypes(), array_shapes(max_dims=2))
+    except ImportError:
+        pass
+    known_type_strats.update(lookup or {})
+    return known_type_strats
+
+
+def get_all_typing_classes():
+    import typing
+    return tuple(
+        cls for cls in (getattr(typing, name) for name in typing.__all__)
+        if isclass(cls) and cls not in (typing.Generic, typing.AbstractSet)
+    ) + (typing.BinaryIO, typing.TextIO)
+
+
+def type_sorting_key(t):
+    """Minimise to None, then non-container types, then container types."""
+    if t is None or t is type(None):
+        return -1
+    return issubclass(t, collections.abc.Container)
+
+
+def resolve_Union(thing):
+    params = getattr(thing, '__union_params__', None) or ()
+    args = getattr(thing, '__args__', None) or ()
+    possible = sorted(params + args, key=type_sorting_key)
+    return st.one_of([from_type(t) for t in possible])
+
+
+def check_by_origin(thing, super_):
+    return thing is super_ or getattr(thing, '__origin__', None) is super_
+
+
+def from_typing_type(thing):
+    import typing
+    # `Any` and `Type` mess up our subclass lookups, so handle them first
+    if thing is typing.Any:
+        from unittest import mock
+        return st.builds(mock.MagicMock)
```

Comment:
I don't like this at all. Currently Hypothesis is very hands-off with mocking. I'm not against changing that in future, but I don't think it should be done casually.

I think we shouldn't infer anything for Any. Generally speaking "Throw  anything at all at this argument" is not an effective approach and people should be specific about what they want.